### PR TITLE
check TB win PV correctness (take 2)

### PR DESCRIPTION
**Pull request by robertnurnberg**
This is the follow-up for https://github.com/vondele/matetrack/pull/95

I did some quick testing, but it probably needs a careful check by a second pair of eyes.

The example below shows that sf-dev has many PVs that lose the win once in TB. This can be seen by status `"draw"`, whereas `"short"` would indicate the PV does not reach TB (which I believe currently is the main concern for @vondele ).

```
> python matecheck.py --engine ./stockfish --epdFile mates2000.epd --syzygyPath /disk2/syzygy/ --nodes 10000
Loaded 2000 FENs, with max(abs(bm)) = 27.

Matetrack started for ./stockfish on mates2000.epd with --nodes 10000 --syzygyPath /disk2/syzygy/ ...
100%|#########################################| 100/100 [00:20<00:00,  4.98it/s]


Using ./stockfish on mates2000.epd with --nodes 10000 --syzygyPath /disk2/syzygy/
Engine ID:     Stockfish dev-20240615-2046c92a
Total FENs:    2000
Found mates:   391
Best mates:    328
Found TB score -20000 with PV status "draw" for FEN "8/8/1N6/8/8/k1K5/1Bp5/1b6 b - -" with bm #-6.
PV: a3a2 b6d5 c2c1q b2c1
Found TB score 20000 with PV status "draw" for FEN "7k/8/5rRK/5P2/8/1B6/8/8 w - -" with bm #8.
PV: h6g5 f6g6 g5g6
Found TB score 20000 with PV status "draw" for FEN "8/8/8/5K2/7k/7p/7N/7N w - -" with bm #9.
PV: h2f3 h4h5 h1g3 h5h6
Found TB score 20000 with PV status "draw" for FEN "8/1B6/5N2/8/b7/8/p7/k1K5 w - -" with bm #7.
PV: f6e4 a4b3 b7a8 b3c2 c1c2
Found TB score -20000 with PV status "draw" for FEN "rr5k/RR6/7K/8/8/8/8/8 b - -" with bm #-6.
PV: h8g8 b7b8 a8b8
Found TB score -20000 with PV status "draw" for FEN "1b6/2p5/8/2K5/k7/8/1PB5/8 b - -" with bm #-7.
PV: a4a5 b2b4 a5a6 c2d3 a6b7
Found TB score 20000 with PV status "draw" for FEN "8/8/8/7K/6Np/7p/7N/7k w - -" with bm #10.
PV: h5g5 h1g1 g5f5 g1h1 f5f4 h1g2 f4e4 g2g3 e4e5 g3g2 e5f4 g2g1 f4f3 g1h1 f3f2
Found TB score 19995 with PV status "draw" for FEN "8/8/4pN2/6p1/5p2/7p/8/5K1k w - -" with bm #9.
PV: f6g4 f4f3 f1f2 e6e5 g4e5 h1h2 e5f3 h2h1
Found TB score -20000 with PV status "wrong" for FEN "4N3/8/8/3p1p2/8/8/p7/k1K5 b - -" with bm #-6.
PV: f5f4 e8g7 f4f3 g7f5
Found TB score 20000 with PV status "draw" for FEN "8/7n/7P/8/6nK/5k2/8/8 b - -" with bm #10.
PV: f3f4 h4h3 h7g5 h3g2
Found TB score -19992 with PV status "draw" for FEN "8/8/8/8/4N3/p1K1N3/p7/rk6 b - -" with bm #-10.
PV: b1c1 e4c5 c1b1 c3d2 b1b2 e3c4 b2b1 c4a3 b1b2 a3c4 b2b1 d2c3 b1c1 c5d3 c1d1
Found TB score -20000 with PV status "draw" for FEN "8/8/8/8/1N5B/p7/4r3/k1K5 b - -" with bm #-8.
PV: e2e6 b4c2 a1a2
Found TB score -19986 with PV status "wrong" for FEN "4R1B1/6pk/8/6K1/8/1p6/p1bp3p/8 b - -" with bm #-9.
PV: h7h8 g8b3 h8h7 b3g8 h7h8 g8a2 h8h7 a2g8 h7h8 g8b3 h8h7 b3c2 g7g6 c2g6 h7g7 e8e7 g7g8 g6h7 g8f8
Found TB score -19992 with PV status "draw" for FEN "8/8/2N5/7K/4N3/6bk/8/5B1n b - -" with bm #-8.
PV: h3h2 c6d4 h2g1 f1h3 g1h2 h5g4 h1f2 e4f2 h2g1 g4g3
Found TB score -19996 with PV status "draw" for FEN "2N5/5K2/4p2k/8/2p3PP/8/8/8 b - -" with bm #-7.
PV: h6h7 c8e7 h7h8 f7e6 h8h7 g4g5 h7g7 g5g6 g7h6 e7g8 h6g6
Found TB score -20000 with PV status "draw" for FEN "1b6/k1p5/2K2N2/8/8/8/8/5B2 b - -" with bm #-8.
PV: a7a8 f6d7 b8a7 c6c7
Found TB score -20000 with PV status "draw" for FEN "8/p7/8/8/P7/3N4/5K1p/7k b - -" with bm #-6.
PV: a7a5 d3f4
Found TB score -20000 with PV status "draw" for FEN "7k/8/4B1K1/8/4R3/1r6/8/8 b - -" with bm #-6.
PV: b3b4 e6g4 b4e4
Found TB score 19999 with PV status "draw" for FEN "5K1k/5PN1/7q/4Q3/8/6p1/8/8 w - -" with bm #7.
PV: e5g3 h6h7 g3h3 h7h3
Found TB score -20000 with PV status "draw" for FEN "8/8/8/8/2Q5/1qN5/k7/2K5 b - -" with bm #-7.
PV: a2a3 c3b1 b3b1 c1b1
Found TB score -20000 with PV status "draw" for FEN "8/8/N7/2B5/8/7p/5K1P/7k b - -" with bm #-7.
PV: h1h2 a6b4
Found TB score -19996 with PV status "draw" for FEN "8/8/8/2Kp4/k1p5/1rp5/8/Q7 b - -" with bm #-14.
PV: b3a3 a1d1 c3c2 d1c2 a3b3 c5d5 a4b4 c2c4 b4a3 c4c1 a3b4 c1c5 b4a4 d5c4
Found TB score -20000 with PV status "draw" for FEN "Q7/8/8/8/8/8/6rp/4K2k b - -" with bm #-11.
PV: h1g1 a8a7 g1h1 a7a6 g2g3 e1f2 g3g2 f2f3
Found TB score -20000 with PV status "draw" for FEN "8/8/8/8/6NK/7p/7N/7k b - -" with bm #-9.
PV: h1g1 h4h3
Found TB score 19997 with PV status "draw" for FEN "8/8/8/3K4/p7/pk1B4/p7/B7 w - -" with bm #11.
PV: d3c4 b3c2 c4a2 c2d3 a1g7 d3d2 g7f8 d2c1 d5c4
Found TB score 19993 with PV status "wrong" for FEN "k7/q2N4/3K4/8/8/6Q1/2p5/n7 w - -" with bm #13.
PV: g3g2 a7b7 g2g8 a8a7 g8g1 a7a6 g1a1 a6b5 a1g1
Found TB score 19991 with PV status "draw" for FEN "6b1/2B2p2/5K1k/8/6P1/3p4/8/8 w - -" with bm #9.
PV: c7f4 h6h7 g4g5 h7h8 f4e5 h8h7 e5c3 d3d2 c3d2 h7h8 d2b4 h8h7 b4c3
Found TB score -20000 with PV status "draw" for FEN "8/1p6/8/2P5/8/8/p2K4/Bk6 b - -" with bm #-9.
PV: b1a1 d2c1
Found TB score 20000 with PV status "draw" for FEN "7k/8/4K3/5N2/3N4/4p3/8/8 w - -" with bm #9.
PV: d4e2 h8g8 f5e7 g8g7 e2f4
Found TB score -19996 with PV status "draw" for FEN "8/8/8/8/2b1p3/3NB1K1/8/3B3k b - -" with bm #-6.
PV: c4a2 d3f2 h1g1 f2e4 g1f1 e4c3 f1e1 e3f2 e1f1 c3a2
Found TB score 20000 with PV status "draw" for FEN "8/8/6p1/4N3/6N1/8/5K2/7k w - -" with bm #7.
PV: e5f3 g6g5 g4e3
Found TB score -19996 with PV status "draw" for FEN "k7/2P5/b3B3/K5Qq/8/8/8/8 b - -" with bm #-6.
PV: h5e8 g5d8 a8a7 d8e8 a6c8 e8c8
Found TB score 20000 with PV status "draw" for FEN "7Q/8/8/8/8/8/6rp/4K2k w - -" with bm #12.
PV: h8a8 h1g1 a8a7 g1h1 a7f7 g2g1 e1f2 g1g2 f2e3
Found TB score -19988 with PV status "draw" for FEN "n6k/4r3/6K1/5p2/8/7Q/8/b7 b - -" with bm #-11.
PV: h8g8 h3b3 g8f8 b3b8 e7e8 b8d6 e8e7 d6d8 e7e8 d8d7 e8e7 d7f5 f8e8 f5b5 e7d7
Found TB score 20000 with PV status "draw" for FEN "8/8/1N6/8/8/k1K5/2p5/Bb6 w - -" with bm #7.
PV: a1b2 a3a2 b6a4
Found TB score -19992 with PV status "draw" for FEN "Q7/8/2Np4/1p6/1pp5/6q1/8/k1K5 b - -" with bm #-10.
PV: g3a3 a8a3 b4a3 c6b4 c4c3 c1c2 d6d5 b4d5 a1a2 d5c3 a2a1
Found TB score -19996 with PV status "draw" for FEN "8/3p4/8/4B3/8/8/r4p2/3R1K1k b - -" with bm #-8.
PV: a2c2 d1d3 c2c1 f1f2 c1c2 f2f3 c2g2 d3d1 g2g1 d1d7 g1f1 f3g4
Found TB score 19993 with PV status "wrong" for FEN "8/8/K7/8/1p6/1p6/1B5p/5B1k w - -" with bm #12.
PV: b2d4 b3b2 f1d3 h1g2 d3e4 g2g3 d4b2 g3f4 a6b5 f4e4 b5b4 h2h1q
Found TB score -20000 with PV status "draw" for FEN "8/8/8/8/4R3/8/p1K1n3/k7 b - -" with bm #-8.
PV: e2d4 c2c3 d4f3 e4b4 f3e1 b4f4
Found TB score -20000 with PV status "draw" for FEN "8/8/6p1/8/2N3N1/8/5K2/7k b - -" with bm #-6.
PV: g6g5 c4d2
Found TB score -20000 with PV status "draw" for FEN "8/2p5/8/2p5/8/7P/p7/k1K5 b - -" with bm #-7.
PV: c5c4 c1c2 c4c3 c2c1 c7c6
Found TB score -20000 with PV status "draw" for FEN "7K/8/8/1p6/8/3N4/2N5/kB6 b - -" with bm #-11.
PV: a1b1 c2a3 b1a2
Found TB score -20000 with PV status "draw" for FEN "8/8/2Q5/8/8/7p/8/q2N1K1k b - -" with bm #-11.
PV: h1h2 c6d6 h2h1 d6d5 h1h2 d5d2 h2g3
Found TB score -20000 with PV status "draw" for FEN "1k6/8/3N4/1N6/6pK/8/8/7B b - -" with bm #-7.
PV: g4g3 h4g3
Found TB score -19998 with PV status "draw" for FEN "8/8/2p5/1kP5/1N6/1P1N4/1K6/8 b - -" with bm #-9.
PV: b5a5 b4c6 a5b5 c6d4 b5a5 b3b4 a5a4 c5c6
Found TB score -19998 with PV status "draw" for FEN "5K1k/5p1q/5B2/8/5P2/6P1/8/8 b - -" with bm #-9.
PV: h7g7 f6g7 h8h7 f8f7
Found TB score 20000 with PV status "draw" for FEN "8/8/8/8/3N2b1/3P4/5K1p/7k w - -" with bm #6.
PV: d4c6 g4h3 c6d4 h3g4 d4c2
Found TB score -20000 with PV status "draw" for FEN "8/4p3/8/8/8/1N3p1p/5K2/7k b - -" with bm #-15.
PV: e7e6 b3d2 e6e5 d2f1 e5e4 f1g3 h1h2 g3e4 h2h1
Found TB score -19996 with PV status "draw" for FEN "7k/8/5N1K/4B3/8/6pp/8/7b b - -" with bm #-8.
PV: g3g2 e5h2 g2g1r h2g1 h1a8 g1d4
Found TB score 20000 with PV status "draw" for FEN "8/3p4/8/6n1/8/8/2R4p/5K1k w - -" with bm #12.
PV: c2c1 g5e4 c1e1 d7d5 e1d1
Found TB wins: 302

Parsing the engine's full UCI output, the following issues were detected:
Bad PVs:       95   (from 50 FENs)
```

Notice also that some PVs seem to turn a win into a loss (though I have not checked these in detail yet).
```
> python matecheck.py --engine ./stockfish --epdFile mates2000.epd --syzygyPath /disk2/syzygy/ --nodes 10000 | grep status | grep -v draw
100%|#########################################| 100/100 [00:18<00:00,  5.48it/s]
Found TB score -19986 with PV status "wrong" for FEN "4R1B1/6pk/8/6K1/8/1p6/p1bp3p/8 b - -" with bm #-9.
Found TB score -20000 with PV status "wrong" for FEN "4N3/8/8/3p1p2/8/8/p7/k1K5 b - -" with bm #-6.
Found TB score 19993 with PV status "wrong" for FEN "k7/q2N4/3K4/8/8/6Q1/2p5/n7 w - -" with bm #13.
Found TB score 19993 with PV status "wrong" for FEN "8/8/K7/8/1p6/1p6/1B5p/5B1k w - -" with bm #12.
```

### CODE REVIEW

**Comment by robertnurnberg** [comment]
I changed the logic in the output order: thinking the TB PV checks would take long, I first show the numbers for the non-TB lines. That means the line `Found TB wins: 302` now appears after all the warning, so a bit out of place.

For the example above the TB PV checks did not take long at all, so I could revert this and have a consistent order: progress bar, warnings, final stats.

Moreover, if for current sf-dev we want to suppress the `"wrong"` and `"draw"` statuses once the win is lost in TB, I could introduce a command line option for that. Let me know. @vondele

**Comment by vondele** [comment]
interesting, yes, that's basically what I would think would happen. 
I guess you did verify it indeed becomes a draw after entering that PV line, right? 

I also think that having an option like `--fullTBPV` (or whatever) would be useful to have.

**Comment by robertnurnberg** [comment]
Reordered the output, and introduced new command line option to only highlight short TB PVs.
```
> python matecheck.py --engine ./stockfish --epdFile mates2000.epd --syzygyPath /disk2/syzygy/ --nodes 10000 --shortTBPVonly
Loaded 2000 FENs, with max(abs(bm)) = 27.

Matetrack started for ./stockfish on mates2000.epd with --nodes 10000 --syzygyPath /disk2/syzygy/ ...
100%|#########################################| 100/100 [00:16<00:00,  5.96it/s]


Using ./stockfish on mates2000.epd with --nodes 10000 --syzygyPath /disk2/syzygy/
Engine ID:     Stockfish dev-20240623-cc992e5e
Total FENs:    2000
Found mates:   385
Best mates:    321
Found TB wins: 301
```

Will do further tests now and report here.

**Comment by robertnurnberg** [comment]
```
> python matecheck.py --engine ./stockfish --epdFile mates2000.epd --syzygyPath /disk2/syzygy/ --nodes 10000 | grep status | grep -v draw
100%|#########################################| 100/100 [00:16<00:00,  5.94it/s]
Found TB score -19988 with PV status "wrong" for FEN "8/2p5/8/2p5/p3Q3/K7/2r5/1k6 b - -" with bm #-8.
Found TB score -20000 with PV status "wrong" for FEN "4N3/8/8/3p1p2/8/8/p7/k1K5 b - -" with bm #-6.
```

The two wrong ones for sf-dev are:
```
Found TB score -19988 with PV status "wrong" for FEN "8/2p5/8/2p5/p3Q3/K7/2r5/1k6 b - -" with bm #-8.
PV: c7c6 e4h1 c2c1 h1h7 c1c2 h7e4 c5c4 e4e1 c2c1 e1b4 b1c2 b4c4 c2d2 c4f4 d2c2 f4a4 c2d3 a4a6 d3d4 a6a5
Found TB score -20000 with PV status "wrong" for FEN "4N3/8/8/3p1p2/8/8/p7/k1K5 b - -" with bm #-6.
PV: f5f4 e8g7 f4f3 g7f5
```

The latter one on [lichess](https://lichess.org/analysis/standard/4N3/8/8/3p1p2/8/8/p7/k1K5_b_-_-) shows that indeed, after `1..f4 2.Ng7` white has turned the win into a loss.

**Comment by vondele** [comment]
so, one feature request and one potential bug.
* feature: can we support multiple directories, e.g. SF allows for `option.SyzygyPath=/home/vondele/chess/syzygy/3-4-5-6/WDL:/home/vondele/chess/syzygy/3-4-5-6/DTZ`  where the directories are separated by `:` this is useful to store DTZ on spinning disk and WDL on SSD for example. python-chess has `add_directory()` that should allow for initializing with multiple directories.
* potential bug: probe_wdl assumes  the position has been reached directly after a capture or pawn move (i.e. 50m counter at 0). Knowing if a position is winning if this is not the case needs considering the dtz. I think the correct logic from that can be found in sf https://github.com/official-stockfish/Stockfish/blob/cc992e5e4a7110b21f85168bdedad7978edad140/src/syzygy/tbprobe.cpp#L1577-L1643 .

**Comment by robertnurnberg** [comment]
I will look into your comments. For now, as a sanity check, the output for the matetrack usage. Should be same as master (module total number of FENs).
```
Loaded 6556 FENs, with max(abs(bm)) = 126.

Matetrack started for ./stockfish on matetrack.epd with --nodes 1000000 ...
100%|###########################################| 97/97 [22:45<00:00, 14.08s/it]


Using ./stockfish on matetrack.epd with --nodes 1000000
Engine ID:     Stockfish dev-20240623-cc992e5e
Total FENs:    6556
Found mates:   3411
Best mates:    2509
```

**Comment by joergoster** [comment]
Question: does the script support different directories for the TBs?
Running The Huntsman with 10000 nodes gives following error at the end of the run:

> Traceback (most recent call last):
>   File "/home/joster/Testsets/Matetrack/matetrack-tb-pvs-sequential/matecheck.py", line 299, in <module>
>     tb = TB(args.syzygyPath) if args.syzygyPath is not None else None
>   File "/home/joster/Testsets/Matetrack/matetrack-tb-pvs-sequential/matecheck.py", line 9, in __init__
>     self.tb = chess.syzygy.open_tablebase(path)
>   File "/home/joster/.local/lib/python3.9/site-packages/chess/syzygy.py", line 1990, in open_tablebase
>     tables.add_directory(directory, load_wdl=load_wdl, load_dtz=load_dtz)
>   File "/home/joster/.local/lib/python3.9/site-packages/chess/syzygy.py", line 1545, in add_directory
>     for filename in os.listdir(directory):
> FileNotFoundError: [Errno 2] No such file or directory: '/media/joster/EGTB-SSD/syzygy-5men/syzygy:/media/joster/EGTB-SSD/wdl:/media/joster/Platte2/syzygy/dtz'
>

**Comment by robertnurnberg** [comment]
> * feature: can we support multiple directories, e.g. SF allows for `option.SyzygyPath=/home/vondele/chess/syzygy/3-4-5-6/WDL:/home/vondele/chess/syzygy/3-4-5-6/DTZ`  where the directories are separated by `:` this is useful to store DTZ on spinning disk and WDL on SSD for example. python-chess has `add_directory()` that should allow for initializing with multiple directories.

Update pushed, please give this a test. Also @joergoster. :)

> * potential bug: probe_wdl assumes  the position has been reached directly after a capture or pawn move (i.e. 50m counter at 0). Knowing if a position is winning if this is not the case needs considering the dtz. I think the correct logic from that can be found in sf https://github.com/official-stockfish/Stockfish/blob/cc992e5e4a7110b21f85168bdedad7978edad140/src/syzygy/tbprobe.cpp#L1577-L1643 .

I would prefer to leave these delicate probing issues out of the script. My understanding is that currently we may potentially miss wrong PVs, that are drawn due to 50mr, because we only check the wdl value of the EGTB. Is that the fundamental issue? Then my answer would be twofold: (a) also for the mate PVs (in the absence of EGTBs) we may in theory declare a PV as `"ok"` that in fact momentarily loses the win in the PV. (we only check for draws by 3-fold or 50mr for now). And secondly, we do have a separate check for 50mr draws in our standard `pv_status` function, that does not rely on EGTBs, but rather on the 50mr check of python chess.

**Comment by joergoster** [comment]
> Update pushed, please give this a test. Also @joergoster. :)

Thanks! Now runs without error.

Note, on Windows the separator is a semicolon.

File: matecheck.py [hunk 1]
```diff
@@ -124,7 +173,15 @@ def analyze_fens(self, fens):
...
     )
-    parser.add_argument("--syzygyPath", help="path to syzygy EGTBs")
+    parser.add_argument(
+        "--syzygyPath", help="path(s) to syzygy EGTBs, with ':' as separator"
+    )
+    parser.add_argument(
+        "--minTBscore",
+        type=int,
+        help="lowest cp score for a TB win",
+        default=20000 - 246,  # for SF this is TB_CP - MAX_PLY
```
Comment (robertnurnberg) [inline matecheck.py:183]: @vondele Is my assumption that this is the lowest possible TBwin score from SF correct?

Comment (vondele) [reply matecheck.py:183]: I think it is ...

**Comment by robertnurnberg** [comment]
> Note, on Windows the separator is a semicolon.

Thanks, added support for this now. Can you check that it still works for you on Linux?

**Comment by joergoster** [comment]
@robertnurnberg Still runs without problems for me.

File: matecheck.py (hunk 1 above)
Comment (joergoster) [reply matecheck.py:184]: Not exactly. Internally it is VALUE_TB - MAX_PLY. I do not know how this tranlates to centipawns exactly.

**Comment by robertnurnberg** [comment]
Another test with a patched version of the same branch.
```
> python matecheck.py --engine ./sf.se --epdFile mates2000.epd --syzygyPath /disk2/syzygy/ --nodes 10000
Loaded 2000 FENs, with max(abs(bm)) = 27.

Matetrack started for ./sf.se on mates2000.epd with --nodes 10000 --syzygyPath /disk2/syzygy/ ...
100%|#########################################| 100/100 [04:13<00:00,  2.54s/it]


Checking 2185 TB win PVs. This may take some time...

Using ./sf.se on mates2000.epd with --nodes 10000 --syzygyPath /disk2/syzygy/
Engine ID:     Stockfish dev-20240626-c6ec39c7
Total FENs:    2000
Found mates:   385
Best mates:    321
Found TB wins: 301
```

Patch is:
```diff
--- a/src/search.cpp
+++ b/src/search.cpp
@@ -1938,7 +1938,7 @@ void SearchManager::check_time(Search::Worker& worker) {
 void syzygy_extend_pv(const OptionsMap& options, Position& pos, RootMove& rootMove) {
 
     auto t_start      = std::chrono::steady_clock::now();
-    int  moveOverhead = int(options["Move Overhead"]);
+    int  moveOverhead = 1000000; // int(options["Move Overhead"]);
 
     auto time_abort = [&t_start, &moveOverhead]() -> bool {
         auto t_end = std::chrono::steady_clock::now();
```

**Pull request merged**